### PR TITLE
Feature/product detail

### PR DESCRIPTION
**Pull request by sweber704**

### CODE REVIEW

File: src/app/store/product-detail/product-detail.component.html [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                <div *ngFor="let size of product.sizes">
+                  <mat-button-toggle type="button" value='{{ size }}'>{{ size }}</mat-button-toggle>
+                </div>
+              </mat-button-toggle-group>
+            </div>
+          </div>
+        </div>
+        <div  fxLayout="row" fxLayoutAlign="space-between">
+          <div fxLayout="column">
+          <h3>Quantity</h3>
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.html]: this should probably be a label tag with a for attribute identifying the id of the associated input

File: src/app/store/product-detail/product-detail.component.scss [hunk 2]
```diff
@@ -0,0 +1,9 @@
+
+mat-form-field {
+    width: 60px;
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.scss]: do we want to promote these styles to a main stylesheet?

File: src/app/store/product-detail/product-detail.component.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    this.initializeForm();
+    this.storeService.getProduct(this.route.snapshot.params['id']).subscribe(result => {
+      this.product = result;
+      if (this.product.sizes && this.product.sizes.length === 0) {
+        this.shoppingForm.get('size').clearValidators();
+        this.shoppingForm.get('size').updateValueAndValidity();
+      } else {
+        this.shoppingForm.get('size').setValidators(Validators.required);
+      }
+    })
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.ts]: should be consistent with end of line use of semicolons

File: src/app/store/product-detail/product-detail.component.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    this.route.params
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.ts]: is the logic below necessary? it looks like ngOnInit is calling getProduct twice and updating the components product twice?

File: src/app/store/product-detail/product-detail.component.html [hunk 5]
```diff
@@ -0,0 +1,58 @@
+<div fxLayout="row">
+  <mat-card fxFlex="60" fxFlex.md="100" fxFlex.sm="100" fxFlex.xs="100">
+    <mat-spinner *ngIf="loading"></mat-spinner>
+    <div fxLayout="row" fxLayoutAlign="center">
+    <img mat-card-image [src]="product.imageUrl" alt="{{ product.name }}" >
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.html]: this page assumes that product will not be null, would be a good idea to wrap the content in an ngIf = product and add some text to display when ngIf = !product

File: src/app/store/product-detail/product-detail.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    })
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(ProductDetailComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
```
Comment (jkoidahl) [inline src/app/store/product-detail/product-detail.component.spec.ts]: tests for addToCart and ngInit would be good to assert the backend gets called and responses are handled appropriately

File: src/app/core/store/store.service.ts [hunk 7]
```diff
@@ -12,4 +12,15 @@ export class StoreService {
   getProductListing(): Observable<Array<Product>> {
     return this.http.get<Array<Product>>(`${environment.API_URL}/store/products`);
   }
+
+  getProduct(index: number): Observable<Product> {
```
Comment (raymer) [inline src/app/core/store/store.service.ts]: I'd call this `id` instead of `index`

File: src/app/store/product-detail/product-detail.component.ts (hunk 4 above)
Comment (raymer) [reply src/app/store/product-detail/product-detail.component.ts]: Yeah not quite sure what's going on here. Double calls to getProduct

File: src/app/core/store/store.service.ts [hunk 8]
```diff
@@ -12,4 +12,15 @@ export class StoreService {
...
+    return this.http.get<Product>(`${environment.API_URL}/store/products/` + index);
+  }
+
+  addToCart(productId: number, size: string, quantity: number) {
```
Comment (raymer) [inline src/app/core/store/store.service.ts]: instead of passing all of these params, I would have a addToCartModel that gets passed in.

File: src/app/store/product-detail/product-detail.component.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  addToCart(): void {
+    this.errors = [];
+    if (this.shoppingForm.invalid) {
+      this.errors.push({ errorDescription: 'Invalid data entered' });
+      return;
+    }
+    this.loading = true;
+    const quantity = this.shoppingForm.value.quantity as number;
```
Comment (raymer) [inline src/app/store/product-detail/product-detail.component.ts]: If you create an addToCartModel, you could just cast the form.value as that type.

File: src/app/store/product-detail/product-detail.component.scss (hunk 2 above)
Comment (roblouie) [reply src/app/store/product-detail/product-detail.component.scss]: I would guess not, otherwise that would change the width of every form field in the whole app, which i assume we don't want, only in product detail? But, idk for sure on the design of that.

File: src/app/store/product-detail/product-detail.component.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  errors: any[] = [];
+  loading = false;
+
+  constructor(
+    private storeService: StoreService,
+    private formBuilder: FormBuilder,
+    private route: ActivatedRoute,
+    private router: Router,
+    private snackBar: MatSnackBar,
+  ) {
```
Comment (roblouie) [inline src/app/store/product-detail/product-detail.component.ts]: This format for params is great when there are a lot. 👍

File: src/app/store/product-detail/product-detail.component.ts (hunk 4 above)
Comment (roblouie) [reply src/app/store/product-detail/product-detail.component.ts]: The below logic also has nested observables which is not good. If the below logic is required (which i suspect it is since ngOnInit is only fired once and we need to listen to param changes), then it should be refactored to chain the observables instead of nest them. Nesting observables or promises semi defeats the purpose, because you're right back in "callback hell", where error handling is very difficult and logic is hard to follow.

File: src/app/store/product-detail/product-detail.component.scss (hunk 2 above)
Comment (sweber704) [reply src/app/store/product-detail/product-detail.component.scss]: I wanted this to apply to only the product detail

File: src/app/store/product-detail/product-detail.component.ts (hunk 3 above)
Comment (sweber704) [reply src/app/store/product-detail/product-detail.component.ts]: agree

File: src/app/store/product-detail/product-detail.component.ts (hunk 4 above)
Comment (sweber704) [reply src/app/store/product-detail/product-detail.component.ts]: My original design had the service storing an array of all products so the getProduct method was synchronous, just returning one product from the array. Phil implemented the product list component with the component holding the list of products instead of the service so when  I added getProduct to his version of the service I made it do an async call to the api to get the single product. I should have chained the observables when I did that instead of nest them.
As Rob pointed out, the purpose of the two calls to getProduct is the first loads the correct product when the page is loaded, the second loads the product if the param changes while we are on the detail page, so if someone changes the param in the address bar. I am thinking I should be using a resolver here to make sure the product is retrieved from the api before the page loads so I will look at doing that.

Comment (sweber704) [reply src/app/store/product-detail/product-detail.component.ts]: I got the resolver working so now it subscribes to the service to get the product data and the component subscribes to the data so if it changes the component updates.

File: src/app/store/product-detail/product-detail.component.html (hunk 5 above)
Comment (sweber704) [reply src/app/store/product-detail/product-detail.component.html]: I will display the error message returned from the api if it doesn't return a product.

File: src/app/core/store/store.service.ts (hunk 7 above)
Comment (sweber704) [reply src/app/core/store/store.service.ts]: sure
